FSM

vladjov00 opened the pull request.

vladjov00 committed:
refactor: Added a FiniteStateMachine abstract class that can be instanced for each object that can be in different states

Each subclass must implement on_enter(state) and on_exit(state) functions.

Created a PlayerFSM class that extends FSM, which should control a player's current state, as well as handle transitions to other states (currently, the states are IDLE, WALKING, AIRBORNE, ATTACKING)
At the moment, PlayerFSM is not linked to the Player object.

vladjov00 committed:
refactor: Players now change their states when doing actions

Removed is_grounded bool variable from Player component - replaced by an instance of PlayerFSM.
Being on the ground is now regulated through states: if the player's state is not AIRBORNE, they are considered to be grounded.

gavrilovmiroslav commented on include/FiniteStateMachine.h:
```diff
@@ -0,0 +1,40 @@
+
+private:
+    // if a pair (t, s) is in the map, performing the transition t from the current state leads to state s
+    containers::Map< Transition, State > newState;
```
No spacing between `<` and `>` and words

gavrilovmiroslav commented on src/PlayerControls.cpp:
```diff
@@ -6,25 +6,23 @@ void PlayerControlsSystem::on_tick()
         {
             player.vertical_velocity = -JUMP_MOD;
-            player.is_grounded = false;
+            player.fsm.change_state(PlayerTransition::JUMP);
```
Looking at this use, I'd recommend you change the name of `change_state` to `trigger` or `fire`. `change_state` suggests that a change of state will happen, and what we want is something that will trigger a change if possible, or do nothing.

gavrilovmiroslav commented on src/PlayerControls.cpp:
```diff
@@ -35,6 +33,7 @@ void PlayerControlsSystem::on_tick()
         if (keys.is_down(bindings.right))
         {
+            player.fsm.change_state(PlayerTransition::WALK);
             replace_component<Flip>(player_entity, None);
```
It just occurred to me: All of this code about what to do with the player could be inside the `on_enter` and `on_exit` functions, otherwise, what are those doing?

gavrilovmiroslav commented on src/PlayerFSM.cpp:
```diff
@@ -0,0 +1,42 @@
+    add_valid_transitions_from_state( PlayerState::ATTACKING, 
+        containers::Set<PlayerTransition> { PlayerTransition::STOP });
+    add_valid_transitions_from_state( PlayerState::AIRBORNE, 
+        containers::Set<PlayerTransition> { PlayerTransition::ATTACK, PlayerTransition::STOP });
```
How do you know which transitions from states correspond to which transitions to states? I'm not sure that this implementation is exactly what you want, and also, it's more complicated than it needs to be, isn't it? Why not just have a map of states and their maps of transitions to new states?

gavrilovmiroslav commented on src/PlayerFSM.cpp:
```diff
@@ -0,0 +1,42 @@
+
+void PlayerFSM::on_exit(PlayerState fromState)
+{
+    std::cout << "exiting state: [" << stateName(currentState) << "], ";
```
Yeah, empty, thought so; maybe we could actually use these? :D

vladjov00 committed:
refactor: PlayerFSM - on_enter() function now changes sprite animation of the player

vladjov00 committed:
feature: Added an airborne sprite animation for wizards, wizards are now different colors

PlayerFSM now holds information of a wizard's color, which is purple for player one and green for player 2.

Commit:
quickfix: fixed animation freezing when collecting pickup

Instead on destorying the entity of the collected pickup, we just delete
it's visibility and pickup components.

vladjov00 committed:
Merged quickfix - resolved conflicts
